packages/zoe/src/contractFacet/zcfZygote.js

```diff
@@ -152,31 +153,64 @@ export const makeZCFZygote = (
           zcfSeat = makeEmptySeatKit().zcfSeat;
         }
         const totalToMint = Object.values(gains).reduce(add, empty);
-        zcfSeat.incrementBy(gains);
+        assert(
+          !zcfSeat.hasExited(),
+          `zcfSeat must be active to mint gains for the zcfSeat`,
+        );
+        const allocationPlusGains = addToAllocation(
+          zcfSeat.getCurrentAllocation(),
+          gains,
+        );
+
+        // Increment the stagedAllocation if it exists so that the
+        // stagedAllocation is kept up to the currentAllocation
+        if (zcfSeat.hasStagedAllocation()) {
+          zcfSeat.incrementBy(gains);
+        }
+
         // verifies offer safety
-        assert(zcfSeat.isOfferSafe(zcfSeat.getStagedAllocation()));
-        // No effects above. The following two steps
-        // *should* be committed atomically, but it is not a
-        // disaster if they are not. If we minted only, no one would
-        // ever get those invisibly-minted assets.
+        assert(
+          zcfSeat.isOfferSafe(allocationPlusGains),
+          `The allocation after minting gains ${allocationPlusGains} for the zcfSeat was not offer safe`,
+        );
+        // No effects above, apart from incrementBy. Note COMMIT POINT within
+        // reallocateForZCFMint. The following two steps *should* be
+        // committed atomically, but it is not a disaster if they are
+        // not. If we minted only, no one would ever get those
+        // invisibly-minted assets.
         E(zoeMintP).mintAndEscrow(totalToMint);
-        // Note COMMIT POINT within reallocateInternal.
-        reallocateInternal([zcfSeat]);
+        reallocateForZCFMint(zcfSeat, allocationPlusGains);
         return zcfSeat;
       },
       burnLosses: (losses, zcfSeat) => {
         assertAmountKeywordRecord(losses, 'losses');
         const totalToBurn = Object.values(losses).reduce(add, empty);
-        zcfSeat.decrementBy(losses);
+        assert(
+          !zcfSeat.hasExited(),
+          `zcfSeat must be active to burn losses from the zcfSeat`,
+        );
+        const allocationMinusLosses = subtractFromAllocation(
+          zcfSeat.getCurrentAllocation(),
+          losses,
+        );
+
+        // Decrement the stagedAllocation if it exists so that the
+        // stagedAllocation is kept up to the currentAllocation
+        if (zcfSeat.hasStagedAllocation()) {
+          zcfSeat.decrementBy(losses);
+        }
```

Comment:
This can fail! It is ok, it fails safe. But a failure here would be surprising, and would often be hard to understand. The contract was only burning seat allocations that it had. It is only a prior staging, which perhaps would not have been used anyway, that has been rendered invalid. One could argue that it is better for the burn to succeed, since it is locally correct, and for the staging to fail to be committed on a later reallocate. On balance I think that's worse, so I am not suggesting it. But it is interesting.

Reply:
Yeah, I agree that would be worse. If the burn succeeded but the staging later failed, it would be even more mysterious to the contract developer.